### DISCUSSION TITLE
Problems with simulations changing when running two stressmodels simultaneously or separate

### DISCUSSION THREAD

**Discussion by LovisaBostrom** [pastas/pastas#296] [Q&A] [answered]
Hello, I am doing a master thesis project where I am using Pastas to simulate the groundwater impact from a tunnel.

I wish to compare how well Pastas can simulate observed groundwater levels depending on if I only use a recharge stress model or also ad leakage as a well model. However, I am getting some weird results where the simulation when using recharge+leakage change if I previously run a simulation with only recharge or not. I am thinking that this should not be the case since the simulations should be independent from each other. 

![image](https://user-images.githubusercontent.com/82801075/115201537-bb048780-a0f5-11eb-9508-47f2261c801b.png)
(First plot is the simulation with only recharge. Second plot is the results after first running only recharge and the recharge+leakage)

![image](https://user-images.githubusercontent.com/82801075/115201698-e9826280-a0f5-11eb-82cc-3894adb70f44.png)
(Simulation when only running the recharge+leakage and NOT previously running a model with only recharge).

![image](https://user-images.githubusercontent.com/82801075/115203104-619d5800-a0f7-11eb-9e1d-1c99de9fb24d.png)
(I also did a decomposition of the time series where it is obvious that also the stresses change between the two different scenarios, which they should not? Left plot: Results from recharge+leakage after first running a model with only recharge. Right plot: recharge+leakage and NOT previously running a model with only recharge)

This is my code for the models if it makes it any clearer:
`# "settings"` 
`ob_func=ps.LmfitSolve`
`rfun=ps.Gamma #responsefunction`
`NO=False #Noise`
`repo=False #report`
`rch=ps.rch.Linear()`
`date_TS='2019-07-01' `
`date_max='2021-04-01' `
`date_min= '2018-10-01'`

`# Model with only recharge `
`ml_WOL=ps.Model(Ho, name=WELL) `
`sm = ps.RechargeModel(rain, evap, rfunc=rfun, recharge=rch, name="rch")`
`ml_WOL.add_stressmodel(sm)`
`ml_WOL.solve( tmin=date_min, report=repo, solver=ob_func, noise=NO)`

`#Model with recharge + leakage`
`m2=ps.Model(Ho, name=WELL)`
`sm2 = ps.RechargeModel(rain, evap, rfunc=rfun, recharge=rch, name="rch")`
`m2.add_stressmodel(sm2)`
`ws2 = ps.WellModel(stresses2, rfunc=ps.HantushWellModel, name="leakage", distances=distances2, settings="well")`
`m2.add_stressmodel(ws2)`
`m2.solve( tmin=date_min,report=repo, solver=ob_func, noise=NO)`

If anyone has any ideas I would be very grateful. Let me know if I should clarify something

Thank you in advance!

**Comment by dbrakenhoff** [comment, marked as answer]
Hi @LovisaBostrom,

Which pastas version are you using? Use `ps.show_versions()` to print the relevant package versions to the console.

If I understand correctly, the results of `m2` are different depending on whether or not you run `ml_WOL` first in the code sample you posted? 

In the decomposition I'm seeing the order of the wells is changed. My thinking is that the order of stresses is changed in subsequent runs (if the kernel is not restarted in between). This is an issue that was fixed in the latest version of pastas (0.17.0) (in commit 41be8c87e22ad961e91ac16adadcfd8818efe262). The old code behavior sorted the stresses by their distances but modified the distances variable in-place, meaning that in a subsequent run, the distances variable was changed, leading to the wrong distances being tied to the wrong stress, which would then lead to a different sorting in a next run if the distances are not reassigned in between. 

So I'm hoping that either updating pastas, or passing a copy of the distances to the WellModel will solve your issue: `... distances=distances2.copy() ...`

Please let me know if that solves your issue!

**Reply by LovisaBostrom** [reply]
Hi @dbrakenhoff ,
I had missed the new update and that resolved the issue!

Thank you so much for your help and fast reply :)

**Reply by LovisaBostrom** [reply]
Hi again @dbrakenhoff,

As I mentioned my previous issue was resolved after updating Pastas. However, when running the model after the update we encountered a new problem with the following error message:

TypeError: loop of ufunc does not support argument 0 of type float which has no callable sqrt method

I am therefore wondering if the update has made the simulation more sensitive to the quality of data? 

What we are trying to do is perform several simulations with different stresses for example only with recharge and then adding leakage and/or infiltration as well models (as in the code I pasted above)

We can resolve this issue by choosing the solver LeastSquares instead of Lmfitsolve! However, since you state on the webpage that the Lmfitsolve is recommended when using a Wellmodel and some of our parameter values are much smaller than 1 we are hesitant in using this approach. Do you have any thoughts?

Thanks in advance!

**Reply by dbrakenhoff** [reply]
Hi @LovisaBostrom, could you open a new thread for this question? I'll answer there. Thanks!

Davíd
